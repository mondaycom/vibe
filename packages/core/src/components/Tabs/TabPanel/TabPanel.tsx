import React, { FC, forwardRef, ReactElement, useRef } from "react";
import cx from "classnames";
import useMergeRef from "../../../hooks/useMergeRef";
import VibeComponentProps from "../../../types/VibeComponentProps";
import { ComponentDefaultTestId, getTestId } from "../../../tests/test-ids-utils";
import styles from "./TabPanel.module.scss";

export interface TabPanelProps extends VibeComponentProps {
  /**
   * The content inside the tab panel.
   */
  children?: ReactElement | ReactElement[] | string;
  /**
   * The index of the tab panel.
   */
  index?: number;
}

const TabPanel: FC<TabPanelProps> = forwardRef(({ className, id, children, index, "data-testid": dataTestId }, ref) => {
  const componentRef = useRef(null);
  const mergedRef = useMergeRef(ref, componentRef);

  return (
    <div
      key={`${id}_${index}`}
      ref={mergedRef}
      className={cx(styles.tabPanelWrapper, className)}
      id={id}
      data-testid={dataTestId || getTestId(ComponentDefaultTestId.TAB_PANEL, id)}
      role="tabpanel"
    >
      {children}
    </div>
  );
});

export default TabPanel;
