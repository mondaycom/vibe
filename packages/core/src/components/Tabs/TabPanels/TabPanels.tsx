import cx from "classnames";
import { camelCase } from "lodash-es";
import React, { FC, forwardRef, ReactElement, useMemo, useRef } from "react";
import useMergeRef from "../../../hooks/useMergeRef";
import VibeComponentProps from "../../../types/VibeComponentProps";
import { TabPanelsAnimationDirection as TabPanelsAnimationDirectionEnum } from "./TabPanelsConstants";
import { TabPanelsAnimationDirection } from "./TabPanels.types";
import { TabPanelProps } from "../TabPanel/TabPanel";
import { ComponentDefaultTestId, getTestId } from "../../../tests/test-ids-utils";
import { getStyle } from "../../../helpers/typesciptCssModulesHelper";
import { withStaticProps } from "../../../types";
import styles from "./TabPanels.module.scss";

export interface TabPanelsProps extends VibeComponentProps {
  /**
   * The index of the currently active tab panel.
   */
  activeTabId?: number;
  /**
   * The animation direction when switching between tab panels.
   */
  animationDirection?: TabPanelsAnimationDirection;
  /**
   * The child elements representing tab panels.
   */
  children?: ReactElement<TabPanelProps> | ReactElement<TabPanelProps>[];
}

const TabPanels: FC<TabPanelsProps> & {
  animationDirections?: typeof TabPanelsAnimationDirectionEnum;
} = forwardRef(
  (
    { className, id, activeTabId = 0, animationDirection = "rtl", children, "data-testid": dataTestId }: TabPanelsProps,
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRef(ref, componentRef);
    const renderedTabs = useMemo(() => {
      return React.Children.map(children, (child, index) => {
        const isActiveTab = activeTabId === index;
        if (!isActiveTab) return null;
        const activeClass = isActiveTab ? "active" : "non-active";
        const animationClass = isActiveTab ? `animation-direction-${animationDirection}` : "";
        return React.cloneElement(child, {
          index,
          ...child.props,
          className: cx(
            styles.tabPanel,
            [getStyle(styles, activeClass)],
            [getStyle(styles, camelCase(animationClass))],
            child.props.className
          )
        });
      }).filter(Boolean);
    }, [children, activeTabId, animationDirection]);

    return (
      <div
        ref={mergedRef}
        className={cx(styles.tabPanelsWrapper, className)}
        id={id}
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.TAB_PANELS, id)}
      >
        {renderedTabs}
      </div>
    );
  }
);

Object.assign(TabPanels, {
  isTabPanels: true
});

export default withStaticProps(TabPanels, {
  animationDirections: TabPanelsAnimationDirectionEnum
});
