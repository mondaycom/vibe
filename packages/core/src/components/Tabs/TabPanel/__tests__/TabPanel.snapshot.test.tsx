import { describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import TabPanels from "../../TabPanels/TabPanels";
import TabPanel from "../TabPanel";

describe("TabPanel renders correctly", () => {
  it("with empty props", () => {
    const tree = renderer
      .create(
        <TabPanels>
          <TabPanel>First</TabPanel>
        </TabPanels>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with className", () => {
    const tree = renderer
      .create(
        <TabPanels>
          <TabPanel className="test">First</TabPanel>
        </TabPanels>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with id", () => {
    const tree = renderer
      .create(
        <TabPanels>
          <TabPanel id="test">First</TabPanel>
        </TabPanels>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
