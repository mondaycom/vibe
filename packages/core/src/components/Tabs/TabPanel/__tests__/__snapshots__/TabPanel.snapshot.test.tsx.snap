// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabPanel renders correctly with className 1`] = `
<div
  className="tabPanelsWrapper"
  data-testid="tab-panels"
>
  <div
    className="tabPanelWrapper tabPanel active animationDirectionRtl test"
    data-testid="tab-panel"
    role="tabpanel"
  >
    First
  </div>
</div>
`;

exports[`TabPanel renders correctly with empty props 1`] = `
<div
  className="tabPanelsWrapper"
  data-testid="tab-panels"
>
  <div
    className="tabPanelWrapper tabPanel active animationDirectionRtl"
    data-testid="tab-panel"
    role="tabpanel"
  >
    First
  </div>
</div>
`;

exports[`TabPanel renders correctly with id 1`] = `
<div
  className="tabPanelsWrapper"
  data-testid="tab-panels"
>
  <div
    className="tabPanelWrapper tabPanel active animationDirectionRtl"
    data-testid="tab-panel_test"
    id="test"
    role="tabpanel"
  >
    First
  </div>
</div>
`;
